src/main/scala/com/rbmhtechnology/calliope/scaladsl/TransactionalEventProducer.scala

```diff
@@ -203,68 +253,140 @@ private class GraphRunnerSupervisor(graph: ProducerGraph, onFailure: FailureHand
 private object ProducerGraphRunner {
 
   case object NotifyCommit
+  case object RunGraph
+  case object StopGraph
+
+  case object GraphStarted
+  case object GraphStopped
 
-  private[ProducerGraphRunner] case object StartGraph
   private[ProducerGraphRunner] case class RestartGraph(failure: Throwable)
 
+  private[ProducerGraphRunner] case class GraphAbortedException() extends RuntimeException
+
+  private[ProducerGraphRunner] sealed trait State
+  private[ProducerGraphRunner] case object Stopped extends State
+  private[ProducerGraphRunner] case object Stopping extends State
+  private[ProducerGraphRunner] case object Running extends State
+
+  private[ProducerGraphRunner] case object Data {
+    def apply(): Data =
+      Data(None, None, Vector.empty)
+  }
+
+  private[ProducerGraphRunner] case class Data(commitQueue: Option[SourceQueueWithComplete[Unit]], killSwitch: Option[UniqueKillSwitch], notifyOnStop: immutable.Seq[ActorRef]) {
+    def withQueue(queue: SourceQueueWithComplete[Unit]): Data =
+      copy(commitQueue = Some(queue))
+
+    def withKillSwitch(killSwitch: UniqueKillSwitch): Data =
+      copy(killSwitch = Some(killSwitch))
+
+    def withNotificationFor(target: ActorRef): Data =
+      copy(notifyOnStop = notifyOnStop :+ target)
+
+    def withoutNotifications(): Data =
+      copy(notifyOnStop = Vector.empty)
+  }
+
   def props(graph: ProducerGraph): Props =
     Props(new ProducerGraphRunner(graph))
 }
 
-private class ProducerGraphRunner(graph: ProducerGraph) extends Actor with ActorLogging with Stash {
+private class ProducerGraphRunner(graph: ProducerGraph) extends Actor
+  with ActorLogging with Stash with FSM[ProducerGraphRunner.State, ProducerGraphRunner.Data] {
 
   import ProducerGraphRunner._
   import context.dispatcher
 
   implicit val materializer = ActorMaterializer()
 
-  private def runGraph(): SourceQueueWithComplete[Unit] = {
-    val (queue, ft) = graph.run()
+  private def runGraph(): (SourceQueueWithComplete[Unit], UniqueKillSwitch) = {
+    val ((queue, ks), ft) = graph.run()
```

Comment:
`queue` and `ks` and never used individually so we could keep them in a single `val`, right? (e.g. `val (graphControl, ft) = graph.run()`